src/utils.cpp

```diff
@@ -360,4 +361,353 @@ int mega_snprintf(char *s, size_t n, const char *format, ...)
+        return 16;
+
+    case AES_CCM_10_08:
+    case AES_GCM_10_08_BROKEN:
+    case AES_GCM_10_08:
+        return 8;
+
+    default:    // unknown block encryption mode
+        return 0;
+    }
+}
+
+unsigned TLVstore::getIvlen(int mode)
+{
+    switch (mode)
+    {
+    case AES_CCM_12_16:
+    case AES_GCM_12_16_BROKEN:
+    case AES_GCM_12_16:
+        return 12;
+
+    case AES_CCM_10_08:
+    case AES_GCM_10_08_BROKEN:
+    case AES_CCM_10_16:
+    case AES_GCM_10_08:
+        return 10;
+
+    default:    // unknown block encryption mode
+        return 0;
+    }
+}
+
+encryptionmode_t TLVstore::getMode(int mode)
+{
+    switch (mode)
+    {
+    case AES_CCM_12_16:
+    case AES_GCM_12_16_BROKEN:
+    case AES_CCM_10_16:
+    case AES_CCM_10_08:
+    case AES_GCM_10_08_BROKEN:
+        return AES_MODE_CCM;
+
+    case AES_GCM_12_16:
+    case AES_GCM_10_08:
+        return AES_MODE_GCM;
+
+    default:    // unknown block encryption mode
+        return AES_MODE_UNKNOWN;
+    }
+}
+
+TLVstore * TLVstore::containerToTLVrecords(const string *data)
+{
+    if (data->empty())
+    {
+        return NULL;
+    }
+
+    TLVstore *tlv = new TLVstore();
+
+    unsigned offset = 0;
+
+    string type;
+    unsigned typelen;
+    string value;
+    unsigned valuelen;
+    size_t pos;
+
+    unsigned datalen = data->length();
+
+    while (offset < datalen)
+    {
+        // get the length of the Type string
+        pos = data->find('\0', offset);
+        typelen = pos - offset;
+
+        // if no valid TLV record in the container, but remaining bytes...
+        if ( (pos == data->npos) || (offset + typelen + 3 > datalen) )
+        {
+            delete tlv;
+            return NULL;
+        }
+
+        // get the Type string
+        type.assign((char*)&(data->data()[offset]), typelen);
+        offset += typelen + 1;        // +1: NULL character
+
+        // get the Length of the value
+        valuelen = data->at(offset) << 8 | data->at(offset + 1);
+        offset += 2;
+
+        // if there's not enough data for value...
+        if (offset + valuelen > datalen)
+        {
+            delete tlv;
+            return NULL;
+        }
+
+        // get the Value
+        value.resize(valuelen);
+        value.assign((char*)&(data->data()[offset]), valuelen);  // value may include NULL characters, read as a buffer
+        offset += valuelen;
+
+        // add it to the map
+        tlv->set(type, value);
+    }
+
+    return tlv;
+}
+
+
+TLVstore * TLVstore::containerToTLVrecords(const string *data, SymmCipher *key)
+{
+    if (data->empty())
+    {
+        return NULL;
+    }
+
+    unsigned offset = 0;
+    encryptionsetting_t encSetting = (encryptionsetting_t) data->at(offset);
+    offset++;
+
+    unsigned ivlen = TLVstore::getIvlen(encSetting);
+    unsigned taglen = TLVstore::getTaglen(encSetting);
+    encryptionmode_t encMode = TLVstore::getMode(encSetting);
+
+    if (!ivlen || !taglen || encMode == AES_MODE_UNKNOWN)
```

Comment:
The code where this checkup should be done is in the `containerToTLVrecords()`, which is the one reading from a buffer. That's already done.
The one pointed by this comment is creating the container from the records, so there's no risk at all.